test/Smart-Account-Factory.specs.ts

```diff
@@ -89,23 +89,200 @@ describe("Smart Account Factory", async () => {
+        await smartAccountFactory.unlockStake(entryPoint.address);
+        // calling unstake() again
+        await expect(
+          smartAccountFactory.unlockStake(entryPoint.address)
+        ).to.be.revertedWith("already unstaking");
       });
-      const depositInfo = await entryPoint.getDepositInfo(
-        smartAccountFactory.address
-      );
-      expect(depositInfo.stake).to.equal(stakeAmount);
     });
 
-    // can unlock
+    describe("withdrawStake(): ", async () => {
+      it("can withdraw Stake", async () => {
+        const { smartAccountFactory, smartAccountImplementation, entryPoint } =
+          await setupTests();
+        // staking first
+        const stakeAmount = ethers.utils.parseEther("1.234256");
+        await smartAccountFactory.addStake(entryPoint.address, 1, {
+          value: stakeAmount,
+        });
+        // waiting for additional blocks
+        const blocksToWait = 10;
+        const currentBlockNumber = await ethers.provider.getBlockNumber();
+        const targetBlockNumber = currentBlockNumber + blocksToWait;
+        while ((await ethers.provider.getBlockNumber()) < targetBlockNumber) {
+          await ethers.provider.send("evm_mine", []);
+        }
+        // withdrawing
+        await smartAccountFactory.unlockStake(entryPoint.address);
+        const tx = await smartAccountFactory.withdrawStake(
+          entryPoint.address,
+          alice.address
+        );
+        await expect(tx).to.emit(entryPoint, "StakeWithdrawn");
+      });
+
+      it("reverts on empty stake withdraw", async () => {
+        const { smartAccountFactory, smartAccountImplementation, entryPoint } =
+          await setupTests();
+        // withdrawing
+        await expect(
+          smartAccountFactory.withdrawStake(entryPoint.address, alice.address)
+        ).to.be.revertedWith("No stake to withdraw");
+      });
 
-    // can withdraw
+      it("reverts when not calling unlockWithdraw() first", async () => {
+        const { smartAccountFactory, smartAccountImplementation, entryPoint } =
+          await setupTests();
+        // staking first
+        const stakeAmount = ethers.utils.parseEther("1.234256");
+        await smartAccountFactory.addStake(entryPoint.address, 1, {
+          value: stakeAmount,
+        });
+        // waiting for additional blocks
+        const blocksToWait = 10;
+        const currentBlockNumber = await ethers.provider.getBlockNumber();
+        const targetBlockNumber = currentBlockNumber + blocksToWait;
+        while ((await ethers.provider.getBlockNumber()) < targetBlockNumber) {
+          await ethers.provider.send("evm_mine", []);
+        }
+        // withdrawing
+        await expect(
+          smartAccountFactory.withdrawStake(entryPoint.address, alice.address)
+        ).to.be.revertedWith("must call unlockStake() first");
+      });
 
-    // not owner cannot add, unlock, withdraw
+      it("reverts when calling before stake withdrawal", async () => {
```

Comment:
what is this test about?

Reply:
It is for [this](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/core/StakeManager.sol#L120) require check in StakeManager.sol. 
Before calling withdrawStake(), it checks whether the withdrawTime is due or not.